Changes to managed_login_version for aws_cognito_user_pool_domain don't work for custom domains

### Terraform and AWS Provider Version

```console
Terraform v1.10.5
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v5.95.0
+ provider registry.terraform.io/hashicorp/awscc v1.46.0
```

### Affected Resource(s) or Data Source(s)

* `aws_cognito_user_pool_domain`

### Expected Behavior

Updating the `managed_login_version` of an existing Cognito custom domain from 1 to 2 should work.

### Actual Behavior

Updating the `managed_login_version` of an existing Cognito custom domain from 1 to 2 doesn't work correctly - the api call fails and the version remains as before the terraform apply.

### Relevant Error/Panic Output

```console
╷
│ Error: updating Cognito User Pool Domain (example.com): operation error Cognito Identity Provider: UpdateUserPoolDomain, https response error StatusCode: 400, RequestID: <redacted>, InvalidParameterException: Cannot convert custom domain to prefix domain.
│ 
│   with aws_cognito_user_pool_domain.customer_subdomain,
│   on main.tf line 136, in resource "aws_cognito_user_pool_domain" "customer_subdomain":
│  136: resource "aws_cognito_user_pool_domain" "customer_subdomain" {
│ 
╵

```


### Sample Terraform Configuration

<details>
<summary>Click to expand configuration</summary>

```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 5.95.0"
    }
  }
}

provider "aws" {
  region = "eu-west-1"
}

provider "aws" {
  alias  = "us_east_1"
  region = "us-east-1"
}

locals {
  use_managed_login = true
  customer_id       = "someid"
  domain            = "example.com"
  hosted_zone_id    = "<HOSTED_ZONE>"
  callback_url      = "https://${local.domain}/auth/sign-in-oidc"
  logout_url        = "https://${local.domain}/logout"
}

resource "aws_cognito_user_pool" "customer_pool" {
  name = local.customer_id

  mfa_configuration = "OPTIONAL"
  user_pool_tier    = "ESSENTIALS"

  software_token_mfa_configuration {
    enabled = true
  }

  deletion_protection = "INACTIVE"

  account_recovery_setting {
    recovery_mechanism {
      name     = "verified_email"
      priority = 1
    }
  }

  username_configuration {
    case_sensitive = false
  }

  username_attributes = ["email"]

  admin_create_user_config {
    allow_admin_create_user_only = true
  }

  schema {
    attribute_data_type = "String"
    mutable             = true
    name                = "email"
    required            = true
    string_attribute_constraints {
      min_length = 1
      max_length = 256
    }
  }

  schema {
    attribute_data_type = "String"
    mutable             = true
    name                = "name"
    required            = true
    string_attribute_constraints {
      min_length = 1
      max_length = 256
    }
  }

  password_policy {
    temporary_password_validity_days = 7
    minimum_length                   = 16
  }
}

resource "aws_cognito_user_pool_client" "portal_client" {
  name                                 = "client"
  user_pool_id                         = aws_cognito_user_pool.customer_pool.id
  callback_urls                        = [local.callback_url]
  logout_urls                          = [local.logout_url]
  allowed_oauth_flows_user_pool_client = true
  allowed_oauth_flows                  = ["code"]
  allowed_oauth_scopes                 = ["email", "aws.cognito.signin.user.admin"]
  supported_identity_providers         = ["COGNITO"]
  explicit_auth_flows                  = ["ALLOW_USER_PASSWORD_AUTH", "ALLOW_REFRESH_TOKEN_AUTH"]
  enable_token_revocation              = true
  prevent_user_existence_errors        = "ENABLED"
  generate_secret                      = true
  access_token_validity                = 5
  id_token_validity                    = 5
  refresh_token_validity               = 60
  token_validity_units {
    access_token  = "minutes"
    id_token      = "minutes"
    refresh_token = "minutes"
  }
}

resource "aws_cognito_user_pool_domain" "amazon_domain" {
  domain                = "some-domain"
  user_pool_id          = aws_cognito_user_pool.customer_pool.id
  managed_login_version = local.use_managed_login ? 2 : 1
}

resource "aws_cognito_user_pool_domain" "customer_subdomain" {
  domain                = local.domain
  certificate_arn       = aws_acm_certificate_validation.this.certificate_arn
  user_pool_id          = aws_cognito_user_pool.customer_pool.id
  managed_login_version = local.use_managed_login ? 2 : 1
}

resource "aws_acm_certificate" "this" {
  provider          = aws.us_east_1
  domain_name       = local.domain
  validation_method = "DNS"

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_route53_record" "this" {
  provider = aws.us_east_1
  for_each = {
    for dvo in aws_acm_certificate.this.domain_validation_options : dvo.domain_name => {
      name   = dvo.resource_record_name
      record = dvo.resource_record_value
      type   = dvo.resource_record_type
    }
  }

  allow_overwrite = true
  name            = each.value.name
  records         = [each.value.record]
  ttl             = 60
  type            = each.value.type
  zone_id         = local.hosted_zone_id
}

resource "aws_acm_certificate_validation" "this" {
  provider                = aws.us_east_1
  certificate_arn         = aws_acm_certificate.this.arn
  validation_record_fqdns = [for record in aws_route53_record.this : record.fqdn]
}
```
</details>


### Steps to Reproduce

1. Use a domain you own and insert the domain and the hosted zone in the sample terraform
2. Apply the configuration
3. Change the managed login version to 2
4. Apply the configuration

### Debug Logging

<details>
<summary>Click to expand log output</summary>

```console
2025-07-02T14:34:13.485+0100 [DEBUG] aws_cognito_user_pool_domain.customer_subdomain: applying the planned Update change
2025-07-02T14:34:13.486+0100 [DEBUG] provider.terraform-provider-aws_v5.95.0_x5: HTTP Request Sent: aws.region=eu-west-1 tf_mux_provider="*schema.GRPCProviderServer" http.method=POST http.request.header.amz_sdk_request="attempt=1; max=25" http.request.header.x_amz_date=20250702T133413Z http.request.header.x_amz_security_token="*****" net.peer.name=cognito-idp.eu-west-1.amazonaws.com tf_resource_type=aws_cognito_user_pool_domain http.request.header.content_type=application/x-amz-json-1.1 @caller=github.com/hashicorp/aws-sdk-go-base/v2@v2.0.0-beta.64/logging/tf_logger.go:45 tf_aws.signing_region="" tf_provider_addr=registry.terraform.io/hashicorp/aws
  http.request.body=
  | {"Domain":"example.com","ManagedLoginVersion":2,"UserPoolId":"<redacted>"}
   http.url=https://cognito-idp.eu-west-1.amazonaws.com/ rpc.method=UpdateUserPoolDomain http.request.header.amz_sdk_invocation_id=<redacted> http.request.header.x_amz_target=AWSCognitoIdentityProviderService.UpdateUserPoolDomain rpc.service="Cognito Identity Provider" rpc.system=aws-api tf_aws.sdk=aws-sdk-go-v2 tf_req_id=<redacted> @module=aws http.request.header.authorization="AWS4-HMAC-SHA256 Credential=<redacted>, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-date;x-amz-security-token;x-amz-target, Signature=*****" http.request_content_length=106 http.user_agent="APN/1.0 HashiCorp/1.0 Terraform/1.10.5 (+https://www.terraform.io) terraform-provider-aws/5.95.0 (+https://registry.terraform.io/providers/hashicorp/aws) aws-sdk-go-v2/1.36.3 ua/2.1 os/macos lang/go#1.23.8 md/GOOS#darwin md/GOARCH#arm64 api/cognitoidentityprovider#1.51.4 m/t,u" tf_rpc=ApplyResourceChange timestamp="2025-07-02T14:34:13.486+0100"
2025-07-02T14:34:13.635+0100 [DEBUG] provider.terraform-provider-aws_v5.95.0_x5: HTTP Response Received: http.response.header.x_amzn_errormessage="Cannot convert custom domain to prefix domain." http.response.header.x_amzn_requestid=<redacted> tf_resource_type=aws_cognito_user_pool_domain http.duration=148 http.response.header.x_amzn_errortype=InvalidParameterException: tf_req_id=<redacted> tf_rpc=ApplyResourceChange tf_mux_provider="*schema.GRPCProviderServer" @caller=github.com/hashicorp/aws-sdk-go-base/v2@v2.0.0-beta.64/logging/tf_logger.go:45 rpc.method=UpdateUserPoolDomain rpc.system=aws-api tf_aws.sdk=aws-sdk-go-v2 rpc.service="Cognito Identity Provider" @module=aws aws.region=eu-west-1
  http.response.body=
  | {"__type":"InvalidParameterException","message":"Cannot convert custom domain to prefix domain."}
   http.response.header.content_type=application/x-amz-json-1.1 http.response.header.date="Wed, 02 Jul 2025 13:34:13 GMT" http.response_content_length=97 http.status_code=400 tf_aws.signing_region="" tf_provider_addr=registry.terraform.io/hashicorp/aws timestamp="2025-07-02T14:34:13.635+0100"
2025-07-02T14:34:13.635+0100 [DEBUG] provider.terraform-provider-aws_v5.95.0_x5: request failed with unretryable error https response error StatusCode: 400, RequestID: 947f9b7e-0bd7-473b-a186-cb8dc0ea4742, InvalidParameterException: Cannot convert custom domain to prefix domain.: rpc.service="Cognito Identity Provider" tf_resource_type=aws_cognito_user_pool_domain @module=aws rpc.system=aws-api tf_aws.sdk=aws-sdk-go-v2 tf_mux_provider="*schema.GRPCProviderServer" tf_provider_addr=registry.terraform.io/hashicorp/aws @caller=github.com/hashicorp/aws-sdk-go-base/v2@v2.0.0-beta.64/logging/tf_logger.go:45 tf_req_id=<redacted> tf_rpc=ApplyResourceChange aws.region=eu-west-1 rpc.method=UpdateUserPoolDomain timestamp="2025-07-02T14:34:13.635+0100"
2025-07-02T14:34:13.636+0100 [ERROR] provider.terraform-provider-aws_v5.95.0_x5: Response contains error diagnostic: tf_proto_version=5.8 tf_provider_addr=registry.terraform.io/hashicorp/aws tf_req_id=<redacted> @caller=github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov5/internal/diag/diagnostics.go:58 @module=sdk.proto diagnostic_detail="" diagnostic_summary="updating Cognito User Pool Domain (example.com): operation error Cognito Identity Provider: UpdateUserPoolDomain, https response error StatusCode: 400, RequestID: <redacted>, InvalidParameterException: Cannot convert custom domain to prefix domain." diagnostic_severity=ERROR tf_resource_type=aws_cognito_user_pool_domain tf_rpc=ApplyResourceChange timestamp="2025-07-02T14:34:13.636+0100"
2025-07-02T14:34:13.648+0100 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2025-07-02T14:34:13.648+0100 [ERROR] vertex "aws_cognito_user_pool_domain.customer_subdomain" error: updating Cognito User Pool Domain (example.com): operation error Cognito Identity Provider: UpdateUserPoolDomain, https response error StatusCode: 400, RequestID: <redacted> InvalidParameterException: Cannot convert custom domain to prefix domain.
╷
│ Error: updating Cognito User Pool Domain (example.com): operation error Cognito Identity Provider: UpdateUserPoolDomain, https response error StatusCode: 400, RequestID: <redacted> InvalidParameterException: Cannot convert custom domain to prefix domain.
│ 
│   with aws_cognito_user_pool_domain.customer_subdomain,
│   on main.tf line 136, in resource "aws_cognito_user_pool_domain" "customer_subdomain":
│  136: resource "aws_cognito_user_pool_domain" "customer_subdomain" {
│ 
╵
2025-07-02T14:34:13.659+0100 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2025-07-02T14:34:13.677+0100 [INFO]  provider: plugin process exited: plugin=.terraform/providers/registry.terraform.io/hashicorp/aws/5.95.0/darwin_arm64/terraform-provider-aws_v5.95.0_x5 id=18257
2025-07-02T14:34:13.677+0100 [DEBUG] provider: plugin exited

```
</details>


### GenAI / LLM Assisted Development

n/a

### Important Facts and References

AWS CLI update to the domain also fails if the certificate is not provided:

https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/create-user-pool-domain.html
https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/update-user-pool-domain.html
https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-add-custom-domain.html

Someone reported the same bug https://github.com/hashicorp/terraform-provider-aws/pull/40855#issuecomment-2846224821

### Would you like to implement a fix?

Yes

--- a/.changelog/43252.txt
+++ b/.changelog/43252.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_cognito_user_pool_domain: Fix managed login version updates for custom Cognito domains
+```

--- a/.changelog/43252.txt
+++ b/.changelog/43252.txt
@@ -1,3 +1,3 @@
 ```release-note:bug
-resource/aws_cognito_user_pool_domain: Fix managed login version updates for custom Cognito domains
+resource/aws_cognito_user_pool_domain: Correctly update `managed_login_version` for custom Cognito domains
 ```

--- a/internal/service/cognitoidp/user_pool_domain.go
+++ b/internal/service/cognitoidp/user_pool_domain.go
@@ -182,6 +182,14 @@ func resourceUserPoolDomainUpdate(ctx context.Context, d *schema.ResourceData, m
 
 	if d.HasChange("managed_login_version") {
 		input.ManagedLoginVersion = aws.Int32(int32(d.Get("managed_login_version").(int)))
+
+		if v, ok := d.GetOk(names.AttrCertificateARN); ok {
+			// If there is a certificate use it as part of the request, this is how AWS distinguishes between custom and prefix domains
+			input.CustomDomainConfig = &awstypes.CustomDomainConfigType{
+				CertificateArn: aws.String(v.(string)),
+			}
+			timeout = 60 * time.Minute // Custom domains take more time to become active.
+		}
 	}
 
 	_, err := conn.UpdateUserPoolDomain(ctx, &input)

--- a/internal/service/cognitoidp/user_pool_domain.go
+++ b/internal/service/cognitoidp/user_pool_domain.go
@@ -184,11 +184,11 @@ func resourceUserPoolDomainUpdate(ctx context.Context, d *schema.ResourceData, m
 		input.ManagedLoginVersion = aws.Int32(int32(d.Get("managed_login_version").(int)))
 
 		if v, ok := d.GetOk(names.AttrCertificateARN); ok {
-			// If there is a certificate use it as part of the request, this is how AWS distinguishes between custom and prefix domains
+			// If there is a certificate use it as part of the request, this is how AWS distinguishes between custom and prefix domains.
+			timeout = 60 * time.Minute // Custom domains take more time to become active.
 			input.CustomDomainConfig = &awstypes.CustomDomainConfigType{
 				CertificateArn: aws.String(v.(string)),
 			}
-			timeout = 60 * time.Minute // Custom domains take more time to become active.
 		}
 	}
 
@@ -235,11 +235,11 @@ func resourceUserPoolDomainDelete(ctx context.Context, d *schema.ResourceData, m
 }
 
 func findUserPoolDomain(ctx context.Context, conn *cognitoidentityprovider.Client, domain string) (*awstypes.DomainDescriptionType, error) {
-	input := &cognitoidentityprovider.DescribeUserPoolDomainInput{
+	input := cognitoidentityprovider.DescribeUserPoolDomainInput{
 		Domain: aws.String(domain),
 	}
 
-	output, err := conn.DescribeUserPoolDomain(ctx, input)
+	output, err := conn.DescribeUserPoolDomain(ctx, &input)
 
 	if errs.IsA[*awstypes.ResourceNotFoundException](err) {
 		return nil, &retry.NotFoundError{

--- a/internal/service/cognitoidp/user_pool_domain_test.go
+++ b/internal/service/cognitoidp/user_pool_domain_test.go
@@ -183,6 +183,43 @@ func TestAccCognitoIDPUserPoolDomain_managedLoginVersion(t *testing.T) {
 	})
 }
 
+func TestAccCognitoIDPUserPoolDomain_customCustomDomainManagedLoginVersionUpdate(t *testing.T) {
+	ctx := acctest.Context(t)
+	rootDomain := acctest.ACMCertificateDomainFromEnv(t)
+	domain := acctest.ACMCertificateRandomSubDomain(rootDomain)
+	poolName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	acmCertResourceName := "aws_acm_certificate.test"
+	cognitoPoolResourceName := "aws_cognito_user_pool_domain.test"
+	fmt.Println("In test")
+	fmt.Println(rootDomain)
+	fmt.Println(poolName)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckRegion(t, endpoints.UsEast1RegionID) },
+		ErrorCheck:               acctest.ErrorCheck(t, names.CognitoIDPServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckUserPoolDomainDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccUserPoolDomainConfig_customCustomDomainManagedLoginVersionUpdate(rootDomain, domain, poolName, 1),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckUserPoolDomainExists(ctx, cognitoPoolResourceName),
+					resource.TestCheckResourceAttr(cognitoPoolResourceName, "managed_login_version", "1"),
+					resource.TestCheckResourceAttrPair(cognitoPoolResourceName, names.AttrCertificateARN, acmCertResourceName, names.AttrARN),
+				),
+			},
+			{
+				Config: testAccUserPoolDomainConfig_customCustomDomainManagedLoginVersionUpdate(rootDomain, domain, poolName, 2),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckUserPoolDomainCertMatches(ctx, cognitoPoolResourceName, acmCertResourceName),
+					resource.TestCheckResourceAttr(cognitoPoolResourceName, "managed_login_version", "2"),
+					resource.TestCheckResourceAttrPair(cognitoPoolResourceName, names.AttrCertificateARN, acmCertResourceName, names.AttrARN),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckUserPoolDomainExists(ctx context.Context, n string) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
 		rs, ok := s.RootModule().Resources[n]
@@ -376,4 +413,44 @@ resource "aws_cognito_user_pool" "test" {
   name = %[1]q
 }
 `, rName, version)
+}
+
+func testAccUserPoolDomainConfig_customCustomDomainManagedLoginVersionUpdate(rootDomain, domain, poolName string, version int) string {
+	return fmt.Sprintf(`
+data "aws_route53_zone" "test" {
+  name         = %[1]q
+  private_zone = false
+}
+
+resource "aws_acm_certificate" "test" {
+  domain_name       = %[2]q
+  validation_method = "DNS"
+}
+
+resource "aws_route53_record" "test" {
+  allow_overwrite = true
+  name            = tolist(aws_acm_certificate.test.domain_validation_options)[0].resource_record_name
+  records         = [tolist(aws_acm_certificate.test.domain_validation_options)[0].resource_record_value]
+  ttl             = 60
+  type            = tolist(aws_acm_certificate.test.domain_validation_options)[0].resource_record_type
+  zone_id         = data.aws_route53_zone.test.zone_id
+}
+
+resource "aws_acm_certificate_validation" "test" {
+  certificate_arn         = aws_acm_certificate.test.arn
+  validation_record_fqdns = [aws_route53_record.test.fqdn]
+}
+
+resource "aws_cognito_user_pool" "test" {
+  name = %[3]q
+}
+
+resource "aws_cognito_user_pool_domain" "test" {
+  certificate_arn = aws_acm_certificate_validation.test.certificate_arn
+  domain          = aws_acm_certificate.test.domain_name
+  user_pool_id    = aws_cognito_user_pool.test.id
+
+  managed_login_version = %[4]d
+}
+`, rootDomain, domain, poolName, version)
 }

--- a/internal/service/cognitoidp/user_pool_domain_test.go
+++ b/internal/service/cognitoidp/user_pool_domain_test.go
@@ -190,9 +190,6 @@ func TestAccCognitoIDPUserPoolDomain_customCustomDomainManagedLoginVersionUpdate
 	poolName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
 	acmCertResourceName := "aws_acm_certificate.test"
 	cognitoPoolResourceName := "aws_cognito_user_pool_domain.test"
-	fmt.Println("In test")
-	fmt.Println(rootDomain)
-	fmt.Println(poolName)
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckRegion(t, endpoints.UsEast1RegionID) },

